en/material-system/effect-syntax.md

```diff
@@ -0,0 +1,259 @@
+you have to do your own porting using the language built-in \_\_VERSION__ macro:
+```glsl
+vec4 fragTextureLod (samplerCube tex, vec3 coord, float lod) {
+  #if __VERSION__ < 130
+    #ifdef GL_EXT_shader_texture_lod
+      return textureCubeLodEXT(tex, coord, lod);
+    #else
+      return textureCube(tex, coord, lod); // fallback to bias
+    #endif
+  #else
+    return textureLod(tex, coord, lod);
+  #endif
+}
+```
+The effect compiler will finally split these compile-time constant branches into different output versions.
+
+### About UBO Memory Layout
+First the conclusion, the final rules are, every non-sampler uniform should be specified in UBO blocks, and for every UBO block:
+* there should be no vec3 typed members；
+* for array typed members, size of each element should be no less than a vec4;
+* any member ordering that introduces a padding will be rejected;
+
+These rules will be checked rigorously at effect compile-time, and throws detailed,implicit padding related compile error.
+
+This might sounds overly-strict at first, but for good reasons:<br>
+Firstly, UBO is a much better basic unit to efficiently reuse data, so discrete declaration is no longer an option;<br>
+Secondly, to this day, many platforms, including WebGL 2.0 only support one platform-independent memory layout, namely std140, and it has many restrictions<sup id="a2">[2](#f2)</sup>:
+* All vec3 members will be aligned to vec4：
+```glsl
+uniform ControversialType {
+  vec3 v3_1; // offset 0, length 16 [IMPLICIT PADDING!]
+}; // total of 16 bytes
+```
+* Any array member with element size less than a vec4 is aligned to vec4 element-wise:
+```glsl
+uniform ProblematicArrays {
+  float f4_1[4]; // offset 0, stride 16, length 64 [IMPLICIT PADDING!]
+}; // total of 64 bytes
+```
+* All UBO members are aligned to the size of itself<sup id="a3">[3](#f3)</sup>:
+```glsl
+uniform IncorrectUBOOrder {
+  float f1_1; // offset 0, length 4 (aligned to 4 bytes)
+  vec2 v2; // offset 8, length 8 (aligned to 8 bytes) [IMPLICIT PADDING!]
+  float f1_2; // offset 16, length 4 (aligned to 4 bytes)
+}; // total of 32 bytes*
+uniform CorrectUBOOrder {
+  float f1_1; // offset 0, length 4 (aligned to 4 bytes)
+  float f1_2; // offset 4, length 4 (aligned to 4 bytes)
+  vec2 v2; // offset 8, length 8 (aligned to 8 bytes)
+}; // total of 16 bytes
+```
+
+This means lots of wasted space, and some driver implementation might not completely comform to the standard<sup id="a4">[4](#f4)</sup>, hence all the strict checking procedure to help clear some pretty insidious bugs.<br>
+
+Note that the actual uniform type can differ from the public interfaces the effect exposes to artists and runtime properties. Through [property target](pass-parameter-list.md#Properties) system, every single channel can be manipulated independently, without restriction of the original uniform.
```

Comment:
Note that the actual uniform type can differ from the public interfaces the effect exposes to artists and runtime properties. Through [property target](pass-parameter-list.md#Properties) system, every single channel can be manipulated independently, without restriction of the original uniform.
-->
Note: that the actual uniform type can differ from the public interfaces the effect exposes to artists and runtime properties. Through the [property target](pass-parameter-list.md#Properties) system, every single channel can be manipulated independently, without the restriction of the original uniform.